### PR TITLE
refactor(ci): unify PR test/coverage summary publishing (#281)

### PR DESCRIPTION
**Pull request by TurboCoder13**
## Summary

Unifies language test reusables and standalone coverage on a single **`publish-test-summary`** flag and shared publish workflows. Fixes the Node bug where `coverage-pr-comment: true` with `post-pr-comment: false` posted nothing. Rich coverage tables are used when coverage was collected; pass/fail totals otherwise.

Also renames quality and artifact comment reusables to **`publish-quality-summary`** and **`publish-artifact-report`** for clearer intent.

## Type of Change

- [ ] New feature (composite action, reusable workflow, or utility script)
- [ ] Bug fix
- [ ] Documentation update
- [x] Refactoring (no functional changes)
- [x] CI/CD improvement

## Checklist

- [x] I have tested these changes locally
- [x] I have updated documentation if needed
- [x] Shell scripts pass `shellcheck`
- [x] YAML files pass `yamllint`
- [x] Python code passes `ruff` and `black`

## Breaking Changes

- `post-pr-comment` → `publish-test-summary` on test reusables and `reusable-coverage`
- Removed `coverage-pr-comment` and dual Node poster jobs; use `publish-test-summary` only
- Replaced `reusable-test-pr-comment.yml` / `reusable-coverage-pr-comment.yml` with `reusable-publish-test-summary.yml`
- Renamed `reusable-quality-pr-comment.yml` → `reusable-publish-quality-summary.yml` (caller job `publish-quality-summary`)
- Renamed `reusable-artifact-pr-comment.yml` → `reusable-publish-artifact-report.yml` (`report-file` input)
- `comment-on-failure` → `publish-validation-report`; `comment-on-pr` → `publish-link-check-report`
- Validation artifact `validation-comment` → `validation-report`
- `draft-pr-skip` default now `true` on Python, Node, and Shell test reusables (aligned with Rust)
- Removed `generate-test-comment.sh`, `generate-coverage-pr-comment.sh`

See `docs/workflow-contract.md` (§ Migration: test summary publishing) and `CHANGELOG.md` [Unreleased].

## Closes

- Closes #281

## Testing

- [x] `uv run lintro chk` (full check, zero issues)
- [x] `uv run lintro tst` (BATS unit + integration)
- [ ] Docker builds (not applicable)

## Quality Checks

- [x] Lint/format via lintro
- [x] Tests via lintro tst
- [ ] Manual workflow run in consumer repo (follow-up for Rustume and others)

> [!NOTE]
> ### Unify PR test and coverage summary publishing across CI reusable workflows
> - Renames all `*-pr-comment` reusable workflows, inputs, jobs, and scripts to `publish-*-summary`/`publish-*-report` equivalents across test, coverage, validation, link-check, and quality workflows
> - Introduces [`reusable-publish-test-summary.yml`](.github/workflows/reusable-publish-test-summary.yml) with support for rich coverage tables sourced from a downloaded artifact, replacing the previous plain pass/fail comment
> - Rust, Node, Python, and shell test workflows now upload coverage artifacts (e.g. LCOV) and delegate to the new publish workflow, which renders a coverage table when coverage is enabled
> - [`generate-test-summary.sh`](scripts/ci/actions/generate-test-summary.sh) (renamed from `generate-test-comment.sh`) now reports ✅ PASSED when the job succeeds with zero parsed tests instead of UNKNOWN
> - Fixes [`run-rust-nextest.sh`](scripts/ci/actions/run-rust-nextest.sh) to capture `cargo nextest` exit status via `PIPESTATUS[0]` instead of `tee`'s exit code
> - Behavioral Change: `draft-pr-skip` now defaults to `true` across all language test reusables; coverage summary jobs no longer require a non-empty coverage percentage to publish
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized e5ca09d.</sup>
> 
>
> 

## Summary by CodeRabbit

* **New Features**
  * Test and coverage reporting now publishes richer, consolidated test/coverage summaries for clearer CI feedback.

* **Breaking Changes**
  * `post-pr-comment` inputs replaced by `publish-test-summary`; summary/report artifact names and defaults changed; some legacy PR-comment workflows removed; draft PRs now skip publishing by default.

* **Bug Fixes**
  * Corrected Rust test exit-code capture and improved coverage-upload detection.

* **Documentation**
  * Terminology updated from “PR comments” to “PR summaries and reports” across docs and examples.

<details><summary><h3>Greptile Summary</h3></summary>

This PR unifies PR comment publishing across all language test reusables and the standalone coverage workflow behind a single `publish-test-summary` flag and a new `reusable-publish-test-summary.yml` dispatcher. It also renames quality and artifact comment reusables for clarity, fixes the Rust nextest exit-code capture, and changes `draft-pr-skip` to default `true` across Python, Node, and Shell test reusables.

- **New `reusable-publish-test-summary.yml`**: routes callers to a rich coverage table (when a coverage artifact is available) or to a plain pass/fail summary generated by `generate-test-summary.sh`, with well-guarded mutual-exclusion conditions between both paths.
- **`run-rust-nextest.sh`**: correctly captures `cargo nextest` exit status via `PIPESTATUS[0]` instead of `tee`'s exit code; all language test callers and the coverage workflow are wired to the new dispatcher.
- **Breaking renames**: `post-pr-comment` → `publish-test-summary`, `comment-on-failure` → `publish-validation-report`, `comment-on-pr` → `publish-link-check-report`; artifact names and deleted legacy scripts are documented in `CHANGELOG.md` and `docs/workflow-contract.md`.
</details>

<h3>Confidence Score: 4/5</h3>

Safe to merge after resolving the reusable-coverage.yml hard-fail edge case from prior threads; newly introduced Rust, Python, and shell paths have correct fallback logic.

The dispatch logic in reusable-publish-test-summary.yml is well-designed with mutually exclusive rich-coverage and plain-summary paths. The Node inline-matrix coverage jobs still produce no PR summary when tests fail with coverage enabled (partially addressed in prior threads). The reusable-coverage.yml caller passes a hardcoded non-empty coverage-artifact-name with an empty coverage-file when the coverage job fails, also tracked in prior threads.

reusable-coverage.yml and the Node publish-test-summary-coverage matrix jobs in reusable-test-node.yml / reusable-test-node-custom.yml

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| .github/workflows/reusable-publish-test-summary.yml | New central dispatcher: orchestrates rich-coverage vs. plain-summary paths via well-guarded conditions. |
| .github/workflows/reusable-coverage.yml | Removed coverage-percent != '' guard; hardcoded coverage-artifact-name='coverage-report' still causes Validate step exit 1 when merged-coverage-file is empty (tracked in prior threads). |
| .github/workflows/reusable-rust-test.yml | Adds LCOV upload + delegates to reusable-publish-test-summary; fallback to plain summary when LCOV is absent works correctly. PIPESTATUS[0] fix is correct. |
| .github/workflows/reusable-test-node.yml | publish-test-summary-coverage and publish-test-summary are correctly mutually exclusive via !inputs.coverage; continue-on-error added to download steps. |
| .github/workflows/reusable-test-node-custom.yml | Non-coverage publish-test-summary intentionally omits test counts; generate-test-summary.sh treats JOB_RESULT=success with zero counts as PASSED. draft-pr-skip default changed to true. |
| .github/workflows/reusable-test-python.yml | coverage: true + upload-coverage: false correctly falls back to plain summary. draft-pr-skip default changed to true. |
| scripts/ci/actions/generate-test-summary.sh | New script with correct status determination; always renders Code Coverage table even for non-coverage workflows. |
| scripts/ci/testing/rust/run-rust-nextest.sh | One-line fix: PIPESTATUS[0] correctly captures cargo nextest exit code instead of tee's exit code. |

</details>

</details>

<details><summary><h3>Flowchart</h3></summary>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[Caller workflow<br/>publish-test-summary: true] --> B{coverage<br/>enabled?}
    B -- "Node/custom:<br/>coverage: true" --> C[publish-test-summary-coverage<br/>matrix job]
    B -- "Rust/Python/Shell<br/>coverage: true" --> D[reusable-publish-test-summary.yml]
    B -- "any language<br/>coverage: false" --> D

    C --> C1[Download per-version<br/>node-coverage-test-summary artifact]
    C1 -- success --> C2[post-pr-comment action<br/>posts coverage table]
    C1 -- fail/continue-on-error --> C3[no comment posted]

    D --> E{prebuilt-test-<br/>summary-file?}
    E -- yes --> F[Load prebuilt file]
    E -- no --> G{rich-coverage-comment<br/>AND artifact-name != ''?}

    G -- yes --> H[Validate coverage-file<br/>not empty]
    H -- empty --> FAIL[exit 1]
    H -- valid --> I[Download coverage artifact<br/>continue-on-error: true]
    I -- success --> J[Generate rich<br/>coverage table]
    I -- fail --> K[Generate plain<br/>test run summary]
    G -- no --> K

    J --> L[Publish from coverage table]
    K --> M[Publish from file]
    F --> M
```
</details>

<a href="https://app.greptile.com/api/ide/cursor?prompt=Fix%20the%20following%202%20code%20review%20issues.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%202%0Ascripts%2Fci%2Factions%2Fgenerate-test-summary.sh%3A215-224%0AThe%20%22Code%20Coverage%22%20table%20is%20rendered%20unconditionally%20in%20every%20test%20summary%2C%20even%20for%20workflows%20that%20never%20collect%20coverage%20%28e.g.%2C%20%60reusable-test-shell.yml%60%20with%20kcov%29.%20When%20%60COVERAGE_PROVIDED%3Dfalse%60%20and%20tests%20pass%2C%20users%20see%20%22%E2%9A%A0%EF%B8%8F%20N%2FA%22%20for%20coverage%20and%20%22Unable%20to%20retrieve%20coverage%3A%20coverage%20report%20was%20not%20generated.%22%20%E2%80%94%20a%20message%20that%20reads%20like%20an%20error%20on%20a%20run%20where%20coverage%20was%20never%20configured.%20Guarding%20the%20section%20on%20%60COVERAGE_PROVIDED%60%20avoids%20the%20noise%20for%20non-coverage%20callers.%0A%0A%60%60%60suggestion%0A%24%28if%20%5B%5B%20%22%24COVERAGE_PROVIDED%22%20%3D%3D%20%22true%22%20%5D%5D%3B%20then%20cat%20%3C%3CCOVSECTION%0A%0A%23%23%23%20%F0%9F%93%8A%20Code%20Coverage%0A%0A%7C%20Metric%20%7C%20Value%20%7C%0A%7C--------%7C-------%7C%0A%7C%20**Line%20Coverage**%20%7C%20%24%7BCOVERAGE_DISPLAY%7D%20%24%7BCOVERAGE_EMOJI%7D%20%7C%0A%7C%20**Threshold**%20%7C%20%24%7BCOVERAGE_THRESHOLD%7D%25%20%7C%0A%7C%20**Status**%20%7C%20%24%7BCOVERAGE_STATUS%7D%20%7C%0ACOVSECTION%0Afi%29%0A%0A%23%23%23%20%F0%9F%93%8B%20Coverage%20Details%0A-%20**Generated%3A**%20%24%28date%20-u%20%2B%25Y-%25m-%25d%29%0A%60%60%60%0A%0A%23%23%23%20Issue%202%20of%202%0A.github%2Fworkflows%2Freusable-test-node-custom.yml%3A549-563%0A**%60publish-test-summary-coverage%60%20runs%20even%20when%20%60aggregate-tests%60%20is%20skipped**%0A%0A%60publish-test-summary-coverage%60%20declares%20%60needs%3A%20%5Bprepare%2C%20test%5D%60%20but%20not%20%60needs%3A%20aggregate-tests%60.%20On%20a%20draft%20PR%20with%20%60draft-pr-skip%3A%20true%60%2C%20%60aggregate-tests%60%20is%20skipped%20while%20this%20matrix%20job%20still%20runs.%20Currently%20the%20job's%20%60if%60%20guards%20independently%20replicate%20the%20draft%20check%20so%20there%20is%20no%20correctness%20issue%2C%20but%20if%20future%20work%20makes%20%60publish-test-summary-coverage%60%20consume%20%60aggregate-tests%60%20outputs%20the%20dependency%20would%20need%20to%20be%20added%20explicitly.%0A%0A&pr=282&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCursorDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCursor.svg?v=3">Fix All in Cursor</picture></a> <a href="https://app.greptile.com/ide/claude-code?prompt=Fix%20the%20following%202%20code%20review%20issues.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%202%0Ascripts%2Fci%2Factions%2Fgenerate-test-summary.sh%3A215-224%0AThe%20%22Code%20Coverage%22%20table%20is%20rendered%20unconditionally%20in%20every%20test%20summary%2C%20even%20for%20workflows%20that%20never%20collect%20coverage%20%28e.g.%2C%20%60reusable-test-shell.yml%60%20with%20kcov%29.%20When%20%60COVERAGE_PROVIDED%3Dfalse%60%20and%20tests%20pass%2C%20users%20see%20%22%E2%9A%A0%EF%B8%8F%20N%2FA%22%20for%20coverage%20and%20%22Unable%20to%20retrieve%20coverage%3A%20coverage%20report%20was%20not%20generated.%22%20%E2%80%94%20a%20message%20that%20reads%20like%20an%20error%20on%20a%20run%20where%20coverage%20was%20never%20configured.%20Guarding%20the%20section%20on%20%60COVERAGE_PROVIDED%60%20avoids%20the%20noise%20for%20non-coverage%20callers.%0A%0A%60%60%60suggestion%0A%24%28if%20%5B%5B%20%22%24COVERAGE_PROVIDED%22%20%3D%3D%20%22true%22%20%5D%5D%3B%20then%20cat%20%3C%3CCOVSECTION%0A%0A%23%23%23%20%F0%9F%93%8A%20Code%20Coverage%0A%0A%7C%20Metric%20%7C%20Value%20%7C%0A%7C--------%7C-------%7C%0A%7C%20**Line%20Coverage**%20%7C%20%24%7BCOVERAGE_DISPLAY%7D%20%24%7BCOVERAGE_EMOJI%7D%20%7C%0A%7C%20**Threshold**%20%7C%20%24%7BCOVERAGE_THRESHOLD%7D%25%20%7C%0A%7C%20**Status**%20%7C%20%24%7BCOVERAGE_STATUS%7D%20%7C%0ACOVSECTION%0Afi%29%0A%0A%23%23%23%20%F0%9F%93%8B%20Coverage%20Details%0A-%20**Generated%3A**%20%24%28date%20-u%20%2B%25Y-%25m-%25d%29%0A%60%60%60%0A%0A%23%23%23%20Issue%202%20of%202%0A.github%2Fworkflows%2Freusable-test-node-custom.yml%3A549-563%0A**%60publish-test-summary-coverage%60%20runs%20even%20when%20%60aggregate-tests%60%20is%20skipped**%0A%0A%60publish-test-summary-coverage%60%20declares%20%60needs%3A%20%5Bprepare%2C%20test%5D%60%20but%20not%20%60needs%3A%20aggregate-tests%60.%20On%20a%20draft%20PR%20with%20%60draft-pr-skip%3A%20true%60%2C%20%60aggregate-tests%60%20is%20skipped%20while%20this%20matrix%20job%20still%20runs.%20Currently%20the%20job's%20%60if%60%20guards%20independently%20replicate%20the%20draft%20check%20so%20there%20is%20no%20correctness%20issue%2C%20but%20if%20future%20work%20makes%20%60publish-test-summary-coverage%60%20consume%20%60aggregate-tests%60%20outputs%20the%20dependency%20would%20need%20to%20be%20added%20explicitly.%0A%0A&repo=lgtm-hq%2Flgtm-ci&pr=282&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInClaudeDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInClaude.svg?v=3">Fix All in Claude Code</picture></a> <a href="https://chatgpt.com/codex/deeplink?prompt=IMPORTANT%3A%20Work%20in%20the%20repository%20%22lgtm-hq%2Flgtm-ci%22%20on%20the%20existing%20branch%20%22refactor%2F281-unify-pr-test-coverage-comments%22.%20Checkout%20that%20branch%20%E2%80%94%20do%20NOT%20create%20a%20new%20branch%20or%20open%20a%20new%20PR.%20Push%20your%20changes%20to%20%22refactor%2F281-unify-pr-test-coverage-comments%22.%0A%0AFix%20the%20following%202%20code%20review%20issues.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%202%0Ascripts%2Fci%2Factions%2Fgenerate-test-summary.sh%3A215-224%0AThe%20%22Code%20Coverage%22%20table%20is%20rendered%20unconditionally%20in%20every%20test%20summary%2C%20even%20for%20workflows%20that%20never%20collect%20coverage%20%28e.g.%2C%20%60reusable-test-shell.yml%60%20with%20kcov%29.%20When%20%60COVERAGE_PROVIDED%3Dfalse%60%20and%20tests%20pass%2C%20users%20see%20%22%E2%9A%A0%EF%B8%8F%20N%2FA%22%20for%20coverage%20and%20%22Unable%20to%20retrieve%20coverage%3A%20coverage%20report%20was%20not%20generated.%22%20%E2%80%94%20a%20message%20that%20reads%20like%20an%20error%20on%20a%20run%20where%20coverage%20was%20never%20configured.%20Guarding%20the%20section%20on%20%60COVERAGE_PROVIDED%60%20avoids%20the%20noise%20for%20non-coverage%20callers.%0A%0A%60%60%60suggestion%0A%24%28if%20%5B%5B%20%22%24COVERAGE_PROVIDED%22%20%3D%3D%20%22true%22%20%5D%5D%3B%20then%20cat%20%3C%3CCOVSECTION%0A%0A%23%23%23%20%F0%9F%93%8A%20Code%20Coverage%0A%0A%7C%20Metric%20%7C%20Value%20%7C%0A%7C--------%7C-------%7C%0A%7C%20**Line%20Coverage**%20%7C%20%24%7BCOVERAGE_DISPLAY%7D%20%24%7BCOVERAGE_EMOJI%7D%20%7C%0A%7C%20**Threshold**%20%7C%20%24%7BCOVERAGE_THRESHOLD%7D%25%20%7C%0A%7C%20**Status**%20%7C%20%24%7BCOVERAGE_STATUS%7D%20%7C%0ACOVSECTION%0Afi%29%0A%0A%23%23%23%20%F0%9F%93%8B%20Coverage%20Details%0A-%20**Generated%3A**%20%24%28date%20-u%20%2B%25Y-%25m-%25d%29%0A%60%60%60%0A%0A%23%23%23%20Issue%202%20of%202%0A.github%2Fworkflows%2Freusable-test-node-custom.yml%3A549-563%0A**%60publish-test-summary-coverage%60%20runs%20even%20when%20%60aggregate-tests%60%20is%20skipped**%0A%0A%60publish-test-summary-coverage%60%20declares%20%60needs%3A%20%5Bprepare%2C%20test%5D%60%20but%20not%20%60needs%3A%20aggregate-tests%60.%20On%20a%20draft%20PR%20with%20%60draft-pr-skip%3A%20true%60%2C%20%60aggregate-tests%60%20is%20skipped%20while%20this%20matrix%20job%20still%20runs.%20Currently%20the%20job's%20%60if%60%20guards%20independently%20replicate%20the%20draft%20check%20so%20there%20is%20no%20correctness%20issue%2C%20but%20if%20future%20work%20makes%20%60publish-test-summary-coverage%60%20consume%20%60aggregate-tests%60%20outputs%20the%20dependency%20would%20need%20to%20be%20added%20explicitly.%0A%0A"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCodexDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCodex.svg?v=3">Fix All in Codex</picture></a>

<sub>Reviews (7): Last reviewed commit: ["fix(ci): restore full download-artifact ..."](https://github.com/lgtm-hq/lgtm-ci/commit/e5ca09d0c04952d6d56bae20e2bb9ebd3bb97ad4) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=35583205)</sub>

### CODE REVIEW

File: .github/workflows/reusable-test-node-custom.yml [hunk 1]
```diff
@@ -593,26 +591,48 @@ jobs:
...
+      )
+    uses: ./.github/workflows/reusable-publish-test-summary.yml
+    permissions:
+      contents: read
+      pull-requests: write
+    with:
+      test-suite-name: ${{ inputs.job-name }}
+      comment-marker: ${{ inputs.comment-marker }}
+      job-result: ${{ needs.test.result }}
+      tooling-ref: ${{ inputs.tooling-ref }}
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply .github/workflows/reusable-test-node-custom.yml:641]: Addressed in f116ffd — we are not leaving this as UNKNOWN.

`reusable-test-node-custom` still does not forward `tests-passed` / `tests-failed` / `tests-total` (the custom `test` job has no parsed counts). Instead, `generate-test-summary.sh` now treats `JOB_RESULT=success` with zero parsed counts as **PASSED** (and updates the per-line test status text accordingly), so `coverage: false` + `publish-test-summary: true` no longer shows **UNKNOWN** on green PRs.

Covered by `test_generate_test_summary.bats` (`reports passed when job succeeded without parsed counts`). Counts in the comment body remain 0 when unparsed; overall status is correct.

File: .github/workflows/reusable-test-node.yml [hunk 2]
```diff
@@ -759,34 +752,36 @@ jobs:
...
           allowed-endpoints: ${{ steps.egress.outputs['allowed-endpoints'] }}
-      - name: Download coverage PR comment artifact
-        uses: actions/download-artifact@3e5f45b2cfb9172054b4087a40e8e0b5a5461e7c # v8.0.1
+      - name: Download coverage test summary artifact
+        id: download-coverage
+        continue-on-error: true
+        uses: actions/download-artifact@3e5f45b2cfb9172054b4087a40e0b5a5461e7c # v8.0.1
         with:
-          name: node-coverage-pr-comment-${{ matrix.node-version }}
+          name: node-coverage-test-summary-${{ matrix.node-version }}
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply .github/workflows/reusable-test-node.yml]: Fixed in the next push — restored the full 40-character pin:

`3e5f45b2cfb9172054b4087a40e8e0b5a5461e7c` (was missing `e8`).

Only `reusable-test-node.yml` `publish-test-summary-coverage` → Download coverage test summary artifact had the typo; other download-artifact pins in this PR already matched.
